frontend/lib/resolution/call-init-deinit.cpp

```diff
@@ -94,35 +100,373 @@ struct CallInitDeinit : VarScopeVisitor {
+        parent->deinitedVars.insert(id);
+      }
+    }
+  }
+}
+
+static bool typeNeedsInitDeinitCall(const Type* t) {
+  if (t->isUnknownType() || t->isErroneousType()) {
+    // can't do anything with these
+    return false;
+  } else if (t->isPrimitiveType() || t->isBuiltinType() || t->isCStringType() ||
+             t->isNilType() || t->isNothingType() || t->isVoidType()) {
+    // OK, we can always default initialize primitive numeric types,
+    // and as well we assume that for the non-generic builtin types
+    // e.g. TaskIdType.
+    // No need to resolve anything additional now.
+    return false;
+  } else if (t->isEnumType()) {
+    // OK, can default-initialize enums to first element
+    return false;
+  } else if (t->isFunctionType()) {
+    return false;
+  } else if (auto ct = t->toClassType()) {
+    auto decorator = ct->decorator();
+    // no action needed for 'borrowed' or 'unmanaged'
+    // (these should just default initialized to 'nil',
+    //  so nothing else needs to be resolved)
+    if (decorator.isBorrowed() || decorator.isUnmanaged()) {
+      return false;
+    }
+  }
+
+  return true;
+}
+
+
+void CallInitDeinit::resolveDefaultInit(const VarLikeDecl* ast, RV& rv) {
+
+  ResolvedExpression& varRes = rv.byAst(ast);
+  QualifiedType varType = varRes.type();
+
+  if (varType.isErroneousType()) {
+    // ignore it -- should have already issued an error
+    return;
+  }
+  if (varType.isUnknownKindOrType()) {
+    context->error(ast, "cannot default initialize variable using generic or unknown type");
+    return;
+  }
+  // check genericity
+  auto g = getTypeGenericity(context, varType);
+  if (g != Type::CONCRETE) {
+    context->error(ast, "cannot default initialize variable using generic or unknown type");
+    return;
+  }
+
+  // these will be set below
+  const ClassType* classType = nullptr;
+  const CompositeType* compositeType = nullptr;
+
+  const Type* vt = varType.type();
+
+  if (auto ct = vt->toClassType()) {
+    auto decorator = ct->decorator();
+    // check that the class is a nilable class type
+    if (!decorator.isNilable()) {
+      // TODO: improve this error
+      context->error(ast, "cannot default initialize variable using non-nilable class type");
+      return;
+    }
+  }
+
+  if (!typeNeedsInitDeinitCall(vt)) {
+    // nothing to do here
+  } else if (vt->isTupleType()) {
+    // TODO: probably need to do something here, at least in some cases
+  } else if (auto ct = vt->toClassType()) {
+    auto decorator = ct->decorator();
+    // check that the class is a nilable class type
+    if (!decorator.isNilable()) {
+      // TODO: improve this error
+      context->error(ast, "cannot default initialize variable using non-nilable class type");
+      return;
+    }
+    // no action needed for 'borrowed' or 'unmanaged'
+    // (these should just default initialized to 'nil',
+    //  so nothing else needs to be resolved)
+    if (decorator.isBorrowed() || decorator.isUnmanaged()) {
```

Comment:
Aren't these checked just above via `typeNeedsInitDeinitCall` and line 186?

Reply:
Could we issue the "cannot default initialize non-nilable class type" in `typeNeedsInitDeinitCall`, too? It'll save us practically all of this error logic here.

Reply:
I'd like to keep the error here, just so `typeNeedsInitDeinitCall` doesn't have any side effects, but I've removed the other duplication, thanks for noticing.